spring-rabbit/src/main/java/org/springframework/amqp/rabbit/config/AbstractRabbitListenerContainerFactory.java

```diff
@@ -370,10 +396,18 @@ public C createListenerContainer(RabbitListenerEndpoint endpoint) {
 
 			endpoint.setupListenerContainer(instance);
 		}
-		if (this.beforeSendReplyPostProcessors != null
-				&& instance.getMessageListener() instanceof AbstractAdaptableMessageListener) {
-			((AbstractAdaptableMessageListener) instance.getMessageListener())
-					.setBeforeSendReplyPostProcessors(this.beforeSendReplyPostProcessors);
+		if (instance.getMessageListener() instanceof AbstractAdaptableMessageListener) {
+			AbstractAdaptableMessageListener messageListener = (AbstractAdaptableMessageListener) instance
+					.getMessageListener();
+			if (this.beforeSendReplyPostProcessors != null) {
+				messageListener.setBeforeSendReplyPostProcessors(this.beforeSendReplyPostProcessors);
+			}
+			if (this.retryTemplate != null) {
+				messageListener.setRetryTemplate(this.retryTemplate);
+			}
+			if (this.recoveryCallback != null) {
+				messageListener.setRecoveryCallback(this.recoveryCallback);
```

Comment:
How about to propagate it only `if (this.retryTemplate != null)`?
I believe a `recoveryCallback` is not going to be used when otherwise...

And that might be reflected in the Javadocs as well.